src/app/archive/archive_lib/processor.ml

```diff
@@ -791,23 +823,41 @@ module Block_and_signed_command = struct
           , Option.map ~f:amount_to_int64 receiver_account_creation_fee_paid
           , Option.map created_token ~f:(fun tid ->
                 Unsigned.UInt64.to_int64 (Token_id.to_uint64 tid) )
-          , balances_to_int64s balances )
+          , balances )
       | Failed (failure, balances) ->
           ( "failed"
           , Some (Transaction_status.Failure.to_string failure)
           , None
           , None
           , None
-          , balances_to_int64s balances )
+          , balances )
+    in
+    let add_optional_balance id balance =
+      match balance with
+      | None ->
+          Deferred.Result.return None
+      | Some balance ->
+          let%map balance_id =
+            Balance.add (module Conn) ~public_key_id:id ~balance
+          in
+          Some balance_id
+    in
+    let%bind fee_payer_balance_id =
+      add_optional_balance fee_payer_id fee_payer_balance
+    in
+    let%bind source_balance_id =
+      add_optional_balance source_id source_balance
+    in
+    let%bind receiver_balance_id =
+      add_optional_balance receiver_id receiver_balance
     in
-    (* TODO: Record these with the transaction *)
-    ignore (fee_payer_balance, source_balance, receiver_balance) ;
     Conn.exec
       (Caqti_request.exec
          Caqti_type.(
-           tup3 (tup3 int int int)
+           tup4 (tup3 int int int)
```

Comment:
the large number of tuples make this a bit awkward

it might be better to define a type, and use `hlist` magic to aid encoding/decoding, as for some of the other modules here

Reply:
Great idea, thanks

Reply:
Ha this even exactly matches the type already defined right above